### PR TITLE
changes to variable storage

### PR DESCRIPTION
**Pull request by MG-longshot**
## Major Changes in this PR
this pull set get status for induvial hardware and removes the need for self.lanes.
added lanes array to Units.lanes
re wrote dependent modules to use new structure

## Notes to Code Reviewers
These changes will allow integration of third party (klipperscreen and mainsail)
UNITS, HUBS, STEPPERS, BUFFERS, EXTRUDERS have get_status for accurate UpToDate reporting

## How the changes in this PR are tested

Ran on Voron2.4 continuing further Indepth testing

## PR Checklist: (Checked-off items are either done or do not apply to this PR)
 
- [x] I have performed a self-review of my code
- [ ] CHANGELOG.md is updated (if end-user facing)
- [x] Sent notification to software-design channel requesting review

### CODE REVIEW

**Comment by MG-longshot** [comment]
this pull set get status for induvial hardware and removes the need for self.lanes. 
added lanes array to Units.lanes
re wrote dependent modules to use new structure

File: extras/AFC.py [hunk 1]
```diff
@@ -1029,36 +1028,35 @@ def HexConvert(self,tmp):
...
+                    str[CUR_UNIT.name][CUR_LANE.name]["weight"]=CUR_LANE.weight
+                    str[CUR_UNIT.name][CUR_LANE.name]["runout_lane"]=CUR_LANE.runout_lane
+                    filiment_stat=self.get_filament_status(CUR_LANE).split(':')
+                    str[CUR_UNIT.name][CUR_LANE.name]['filament_status']=filiment_stat[0]
+                    str[CUR_UNIT.name][CUR_LANE.name]['filament_status_led']=filiment_stat[1]
+                    str[CUR_UNIT.name][CUR_LANE.name]['status'] = CUR_LANE.status 
+                    numoflanes +=1
+                str[CUR_UNIT.name]['system']={}
+                str[CUR_UNIT.name]['system']['type'] = CUR_UNIT.type
+                #str[UNIT]['system']['hub_loaded']  = True == CUR_UNIT.hub_loaded
```
Comment (jimmyjon711) [inline extras/AFC.py:1057]: Is this going to be re-added at some point?

Comment (MG-longshot) [reply extras/AFC.py:1057]: Has to be moved haven't decided where yet

File: extras/AFC.py [hunk 2]
```diff
@@ -1067,23 +1065,35 @@ def get_status(self, eventtime):
...
         str["system"]["extruders"]={}
 
-        for EXTRUDE in self.extruders.keys():
-            str["system"]["extruders"][EXTRUDE]={}
-            CUR_EXTRUDER = self.printer.lookup_object('AFC_extruder ' + EXTRUDE)
-            str["system"]["extruders"][EXTRUDE]['lane_loaded'] = CUR_EXTRUDER.lane_loaded
-            if CUR_EXTRUDER.tool_start == "buffer":
-                if CUR_EXTRUDER.lane_loaded == '':
-                    str ["system"]["extruders"][EXTRUDE]['tool_start_sensor'] = False
+        for EXTRUDE in self.printer.objects:
```
Comment (jimmyjon711) [inline extras/AFC.py:1068]: Why did this move away from `self.extruder.keys()`? Iterating over all the printer object every time is going to be costly, compared to an array that is going to be smaller.

Comment (MG-longshot) [reply extras/AFC.py:1068]: The afc get_status going away soon as mainsail adjusted

File: extras/AFC_prep.py [hunk 3]
```diff
@@ -64,95 +62,88 @@ def PREP(self, gcmd):
...
 
         for PO in self.printer.objects:
             if 'AFC_stepper' in PO and 'tmc' not in PO:
                 LANE=self.printer.lookup_object(PO)
-                self.AFC.stepper[LANE.name]=LANE
+                UNIT=self.printer.lookup_object('AFC_unit '+ LANE.unit)
+                if LANE.name not in UNIT.lanes: UNIT.lanes.append(LANE.name)    #add lanes to units list
+                self.AFC.stepper[LANE.name]=LANE                                #add list of all lanes
 
-                # If extruder section exists in vars file add currently stored data to AFC.extruders array
```
Comment (jimmyjon711) [inline extras/AFC_prep.py:73]: With the removal of setting `self.AFC.extruders` when `system_test` is called it errors out. Not sure why some of the stuff was redone in this file.

image

Comment (MG-longshot) [reply extras/AFC_prep.py:73]: Weird indoor ger that error 

Look when I get home

Comment (MG-longshot) [reply extras/AFC_prep.py:73]: Ir was done for all the reasons of the last week

**Pull request closed without merging**